torchmetrics/image/sam.py

```diff
@@ -0,0 +1,89 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, Dict, List, Optional
+
+from torch import Tensor
+from typing_extensions import Literal
+
+from torchmetrics.functional.image.sam import _sam_compute, _sam_update
+from torchmetrics.metric import Metric
+from torchmetrics.utilities import rank_zero_warn
+from torchmetrics.utilities.data import dim_zero_cat
+
+
+class SpectralAngleMapper(Metric):
+    """Spectral Angle Mapper determine thespectral similarity between image spectra and reference spectra by
+    calculating the angle between thespectra, where small angles between the two spectrums indicate high similarity
+    and high angles indicate low similarity.
+
+    Args:
+        reduction: a method to reduce metric score over labels.
+            - ``'elementwise_mean'``: takes the mean (default)
+            - ``'sum'``: takes the sum
+            - ``'none'``: no reduction will be applied
+        compute_on_step:
+            Forward only calls ``update()`` and returns None if this is set to False.
+
+        kwargs:
+            Additional keyword arguments, see :ref:`Metric kwargs` for more info.
+    Return:
+        Tensor with SpectralAngleMapper score
+
+    Example:
+        >>> import torch
+        >>> from torchmetrics import SpectralAngleMapper
+        >>> preds = torch.rand([16, 1, 16, 16], generator=torch.manual_seed(42))
+        >>> target = torch.rand([16, 1, 16, 16], generator=torch.manual_seed(123))
+        >>> sam = SpectralAngleMapper()
+        >>> sam(preds, target)
+        tensor(0.7377)
+    """
+
+    preds: List[Tensor]
+    target: List[Tensor]
+    higher_is_better: bool = False
+
+    def __init__(
+        self,
+        reduction: Literal["elementwise_mean", "sum", "none"] = "elementwise_mean",
+        compute_on_step: Optional[bool] = None,
+        **kwargs: Dict[str, Any],
+    ) -> None:
+        super().__init__(compute_on_step=compute_on_step, **kwargs)
+        rank_zero_warn(
+            "Metric `SpectralAngleMapper` will save all targets and"
+            " predictions in buffer. For large datasets this may lead"
+            " to large memory footprint."
+        )
+
+        self.add_state("preds", default=[], dist_reduce_fx="cat")
+        self.add_state("target", default=[], dist_reduce_fx="cat")
```

Comment:
The `SpectralAngleMapper` metric doesn't have to store all targets and predictions. It just needs to store the sum of the angles and the number of observations (`B * H * W`), as [done for the `MeanSquaredError` metric](https://github.com/PyTorchLightning/metrics/blob/3af729508289d3babf0e166d9e8405cb2b0758a2/torchmetrics/functional/regression/mse.py#L34).

Reply:
I think it's better in the update stage, we only do check the validity of input data (press and target), then we will compute everything at compute stage at a time. It will make more clear to read and maintain as [done for the `CosineSimilarity`](https://github.com/PyTorchLightning/metrics/blob/3af729508289d3babf0e166d9e8405cb2b0758a2/torchmetrics/regression/cosine_similarity.py)

Reply:
I don't know if there are any guidelines on it, but as of now the metric will:

- "have a large memory footprint" -- as quoted from the warning message
- not allow for aggregation on batches with variable H and W dimensions -- from what I understand of the `dist_reduce_fx="cat"` argument given to the `preds` and `target` states.

But as you said that's what's done for the `CosineSimilarity` metric which is closely related to the `SpectralAngleMapper` so I don't know what is better.